diff.py

```diff
@@ -2232,6 +2232,17 @@ def process(dump: str, config: Config) -> List[Line]:
                 break
             i += 1
 
+        is_text_relative_j = False
+        if (
+            arch.name == "mips"
```

Comment:
So do something like
```diff
diff --git a/diff.py b/diff.py
index 1b3309e..5c5aa03 100755
--- a/diff.py
+++ b/diff.py
@@ -1718,6 +1718,7 @@ class ArchSettings:
     re_imm: Pattern[str]
     re_reloc: Pattern[str]
     branch_instructions: Set[str]
+    jump_instructions: Set[str]
     instructions_with_address_immediates: Set[str]
     forbidden: Set[str] = field(default_factory=lambda: set(string.ascii_letters + "_"))
     arch_flags: List[str] = field(default_factory=list)
@@ -1755,6 +1756,19 @@ MIPS_BRANCH_INSTRUCTIONS = MIPS_BRANCH_LIKELY_INSTRUCTIONS.union(
     }
 )
 
+MIPS_JUMP_ADDR_INSTRUCTIONS = {
+    "jal",
+    "j",
+}
+MIPS_JUMP_REG_INSTRUCTIONS = {
+    "jalr",
+    "jr",
+}
+
+MIPS_JUMP_INSTRUCTIONS = set.union(
+    MIPS_JUMP_ADDR_INSTRUCTIONS, MIPS_JUMP_REG_INSTRUCTIONS
+)
+
 ARM32_PREFIXES = {"b", "bl"}
 ARM32_CONDS = {
     "",
@@ -1918,8 +1932,11 @@ MIPS_SETTINGS = ArchSettings(
     arch_flags=["-m", "mips:4300"],
     branch_likely_instructions=MIPS_BRANCH_LIKELY_INSTRUCTIONS,
     branch_instructions=MIPS_BRANCH_INSTRUCTIONS,
-    instructions_with_address_immediates=MIPS_BRANCH_INSTRUCTIONS.union({"jal", "j"}),
-    delay_slot_instructions=MIPS_BRANCH_INSTRUCTIONS.union({"j", "jal", "jr", "jalr"}),
+    jump_instructions=MIPS_JUMP_INSTRUCTIONS,
+    instructions_with_address_immediates=set.union(
+        MIPS_BRANCH_INSTRUCTIONS, MIPS_JUMP_ADDR_INSTRUCTIONS
+    ),
+    delay_slot_instructions=set.union(MIPS_BRANCH_INSTRUCTIONS, MIPS_JUMP_INSTRUCTIONS),
     proc=AsmProcessorMIPS,
 )
 
@@ -1942,6 +1959,7 @@ ARM32_SETTINGS = ArchSettings(
     re_imm=re.compile(r"(?<!sp, )#-?(0x[0-9a-fA-F]+|[0-9]+)\b"),
     re_reloc=re.compile(r"R_ARM_"),
     branch_instructions=ARM32_BRANCH_INSTRUCTIONS,
+    jump_instructions=set(),
     instructions_with_address_immediates=ARM32_BRANCH_INSTRUCTIONS.union({"adr"}),
     proc=AsmProcessorARM32,
 )
@@ -1963,6 +1981,7 @@ AARCH64_SETTINGS = ArchSettings(
     re_imm=re.compile(r"(?<!sp, )#-?(0x[0-9a-fA-F]+|[0-9]+)\b"),
     re_reloc=re.compile(r"R_AARCH64_"),
     branch_instructions=AARCH64_BRANCH_INSTRUCTIONS,
+    jump_instructions=set(),
     instructions_with_address_immediates=AARCH64_BRANCH_INSTRUCTIONS.union(
         {"bl", "adrp"}
     ),
@@ -1983,6 +2002,7 @@ PPC_SETTINGS = ArchSettings(
     re_reloc=re.compile(r"R_PPC_"),
     arch_flags=["-m", "powerpc", "-M", "broadway"],
     branch_instructions=PPC_BRANCH_INSTRUCTIONS,
+    jump_instructions=set(),
     instructions_with_address_immediates=PPC_BRANCH_INSTRUCTIONS.union({"bl"}),
     proc=AsmProcessorPPC,
 )
@@ -2010,6 +2030,7 @@ I686_SETTINGS = ArchSettings(
     # This destroys the objdump output processor logic, so we avoid this.
     arch_flags=["-m", "i386", "--no-show-raw-insn"],
     branch_instructions=I686_BRANCH_INSTRUCTIONS,
+    jump_instructions=set(),
     instructions_with_address_immediates=I686_BRANCH_INSTRUCTIONS.union({"mov"}),
     proc=AsmProcessorI686,
 )
@@ -2234,8 +2255,7 @@ def process(dump: str, config: Config) -> List[Line]:
 
         is_text_relative_j = False
         if (
-            arch.name == "mips"
-            and mnemonic == "j"
+            mnemonic in arch.jump_instructions
             and symbol is not None
             and symbol.startswith(".text")
         ):
```
?

Reply:
putting `jal` in there is an interesting idea, I guess it works though I'd lean against it and prefer to see `.text + ...` in that case without a branch arrow. And I think we should exclude jr/jalr too, no need to change behavior for those. But otherwise yeah. Could use `= default_factory(set)` to avoid the need to define it for other archs.

Definitely keep this MIPS-only for now, other arches can be fixed later if we encounter the same problem there.

Reply:
Oh, did you mean for `jump_instructions` to only contain `j`? I don't think that's much of an improvement over just using the mnemonic.

Reply:
True, I just figured it's a natural way to cover both `mips` and `mipsel`. But extending the arch.name check works as well I suppose. (Apparently we make the same mistake about missing `mipsel` in every other place where arch.name is checked for too.)